src/Dropdown/Dropdown.js

```diff
@@ -97,36 +152,68 @@ export default class Dropdown extends Util.mixin(BindMixin) {
 
   handleMenuToggle(e) {
     e.stopPropagation();
+    let menuDirection = this.state.menuDirection;
+
+    // If the menu isn't open, then we're about to open it and we need to
+    // calculate the direction every time.
+    if (!this.state.isOpen) {
+      menuDirection = this.getMenuDirection(this.state.menuHeight);
+    }
+
     this.setState({
+      menuDirection,
       isOpen: !this.state.isOpen
     });
   }
 
   render() {
+    // Set a key based on the menu height so that React knows to keep the
+    // the DOM element around while we are measuring it.
+    let props = this.props;
+    let state = this.state;
+    let dropdownKey = state.knowMenuHeight || 'initial-render';
+    let dropdownMenu = null;
+    let dropdownMenuClassSet = classNames(
+      state.menuDirection,
+      props.dropdownMenuClassName
+    );
     let dropdownStateClassSet = {
-      'open': this.state.isOpen
+      'open': state.isOpen
     };
-    let dropdownMenu = null;
-    let items = this.props.items;
+    let items = props.items;
+    let transitionName =
+      `${props.transitionName}-${state.menuDirection}`;
     let wrapperClassSet = classNames(
+      state.menuDirection,
       dropdownStateClassSet,
-      this.props.wrapperClassName
+      props.wrapperClassName
     );
 
-    if (this.state.isOpen) {
+    if (state.isOpen) {
       dropdownMenu = (
-        <span className={this.props.dropdownMenuClassName}
-          role="menu">
-          <ul className={this.props.dropdownMenuListClassName}>
-            {this.getMenuItems(items)}
+        <span className={dropdownMenuClassSet}
+          role="menu" ref="dropdownMenu" key={dropdownKey}>
+          <ul className={props.dropdownMenuListClassName}>
+            {this.getMenuItems(props.items)}
           </ul>
         </span>
       );
     }
 
-    if (this.props.transition) {
+    // If we don't know the menu's height, we render it invisibly and then
+    // immediately measure its height in #componentDidUpdate, which will change
+    // the state and trigger another render.
+    if (state.isOpen && state.knowMenuHeight === false) {
```

Comment:
Could we join this `if` with the one above so that if the dropdown is `open` but we don't know the height we don't end up going into two `if` statements.